compiler/rustc_mir/src/transform/copy_propagation.rs

```diff
@@ -0,0 +1,310 @@
+//! A intra-block copy propagation pass.
+//!
+//! Given an assignment `_a = _b` replaces subsequent uses of destination `_a` with source `_b`, as
+//! long as neither `a` nor `_b` had been modified in the intervening statements.
+//!
+//! The implementation processes block statements & terminator in the execution order. For each
+//! local it keeps track of a source that defined its current value. When it encounters a copy use
+//! of a local, it verifies that source had not been modified since the assignment and replaces the
+//! local with the source.
+//!
+//! To detect modifications, each local has a generation number that is increased after each direct
+//! modification. The local generation number is recorded at the time of the assignment and
+//! verified before the propagation to ensure that the local remains unchanged since the
+//! assignment.
+//!
+//! Instead of detecting indirect modifications, locals that have their address taken never
+//! participate in copy propagation.
+//!
+//! When moving in-between the blocks, all recorded values are invalidated. To do that in O(1)
+//! time, generation numbers have a global component that is increased after each block.
+
+use crate::transform::MirPass;
+use crate::util::ever_borrowed_locals;
+use rustc_index::bit_set::BitSet;
+use rustc_index::vec::IndexVec;
+use rustc_middle::mir::visit::*;
+use rustc_middle::mir::*;
+use rustc_middle::ty::{List, TyCtxt};
+
+pub struct CopyPropagation;
+
+impl<'tcx> MirPass<'tcx> for CopyPropagation {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        copy_move_operands_in_statements(tcx, body);
```

Comment:
Should the logic be reversed, to transform moves into copies only for propagated locals?

Reply:
While it would be possible, I intentionally wanted to avoid doing that for two reasons:

* Currently there are no benefits from retaining move operands inside statements.
* During copy propagation, you don't know ahead of time whether operand should be transformed or not. They would have to be transformed after the fact. I didn't want to depend on such a transformation for correctness.